src/System.Private.CoreLib/shared/System/Threading/ThreadPool.cs

```diff
@@ -474,55 +1340,89 @@ public void Enqueue(object callback, bool forceGlobal)
             if (loggingEnabled)
                 System.Diagnostics.Tracing.FrameworkEventSource.Log.ThreadPoolEnqueueWorkObject(callback);
 
-            ThreadPoolWorkQueueThreadLocals? tl = null;
-            if (!forceGlobal)
-                tl = ThreadPoolWorkQueueThreadLocals.threadLocals;
-
-            if (null != tl)
+            var localQueue = GetOrAddLocalQueue();
+            if (forceGlobal)
             {
-                tl.workStealingQueue.LocalPush(callback);
+                _globalQueue.Enqueue(callback, localQueue);
             }
             else
             {
-                workItems.Enqueue(callback);
+                localQueue.Enqueue(callback);
             }
 
-            EnsureThreadRequested();
+            // make sure there is at least one worker request
+            // as long as there is one worker to come, this item will be noticed.
+            // in fact ask anyways, up to #proc
+            RequestThread();
         }
 
-        internal bool LocalFindAndPop(object callback)
+        /// <summary>
+        ///  Task in these states cannot be inlined
+        /// </summary>
+        private const int TASK_STATE_COMPLETED_OR_INVOKED = Task.TASK_STATE_CANCELED | Task.TASK_STATE_FAULTED | Task.TASK_STATE_RAN_TO_COMPLETION | Task.TASK_STATE_DELEGATE_INVOKED;
+
+        internal bool TryRemove(Task callback)
         {
-            ThreadPoolWorkQueueThreadLocals? tl = ThreadPoolWorkQueueThreadLocals.threadLocals;
-            return tl != null && tl.workStealingQueue.LocalFindAndPop(callback);
+            int localQueueIndex = GetLocalQueueIndex();
+            LocalQueue? localQueue = _localQueues[localQueueIndex];
+            if (localQueue != null && localQueue.TryRemove(callback))
+            {
+                return true;
+            }
+
+            for (int i = 0; i < _localQueues.Length; ++i)
```

Comment:
Take `_localQueues` to local to avoid bounds check in loop? https://github.com/dotnet/coreclr/issues/27510

Reply:
We generally do not make this far. The task is either in the local queue or stolen and executed. 
However, there is code in existence that relies on task inlining (or more specifically on task.Wait not blocking), so we check other queues if task did not run yet.

Storing queues in a temp is still better. Thanks!